responsemanager/responseassembler/responseassembler.go

```diff
@@ -70,46 +64,99 @@ func New(ctx context.Context, peerHandler PeerMessageHandler) *ResponseAssembler
+}
+
+func (r *responseStream) Close() error {
+	r.closedLk.Lock()
+	r.closed = true
+	r.closedLk.Unlock()
+	return nil
+}
+
+func (r *responseStream) isClosed() bool {
+	r.closedLk.RLock()
+	defer r.closedLk.RUnlock()
+	return r.closed
+}
+
+type ResponseStream interface {
+	Transaction(transaction Transaction) error
+	DedupKey(key string)
+	IgnoreBlocks(links []ipld.Link)
+	SkipFirstBlocks(skipFirstBlocks int64)
+	// ClearRequest removes all tracking for this request.
+	ClearRequest()
+}
+
 // DedupKey indicates that outgoing blocks should be deduplicated in a seperate bucket (only with requests that share
 // supplied key string)
-func (ra *ResponseAssembler) DedupKey(p peer.ID, requestID graphsync.RequestID, key string) {
-	ra.GetProcess(p).(*peerLinkTracker).DedupKey(requestID, key)
+func (rs *responseStream) DedupKey(key string) {
+	rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).DedupKey(rs.requestID, key)
 }
 
 // IgnoreBlocks indicates that a list of keys should be ignored when sending blocks
-func (ra *ResponseAssembler) IgnoreBlocks(p peer.ID, requestID graphsync.RequestID, links []ipld.Link) {
-	ra.GetProcess(p).(*peerLinkTracker).IgnoreBlocks(requestID, links)
+func (rs *responseStream) IgnoreBlocks(links []ipld.Link) {
+	rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).IgnoreBlocks(rs.requestID, links)
 }
 
 // SkipFirstBlocks tells the assembler for the given request to not send the first N blocks
-func (ra *ResponseAssembler) SkipFirstBlocks(p peer.ID, requestID graphsync.RequestID, skipFirstBlocks int64) {
-	ra.GetProcess(p).(*peerLinkTracker).SkipFirstBlocks(requestID, skipFirstBlocks)
+func (rs *responseStream) SkipFirstBlocks(skipFirstBlocks int64) {
+	rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).SkipFirstBlocks(rs.requestID, skipFirstBlocks)
+}
+
+// ClearRequest removes all tracking for this request.
+func (rs *responseStream) ClearRequest() {
+	_ = rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker).FinishTracking(rs.requestID)
 }
 
-// Transaction builds a response, and queues it for sending in the next outgoing message
-func (ra *ResponseAssembler) Transaction(p peer.ID, requestID graphsync.RequestID, transaction Transaction) error {
+func (rs *responseStream) Transaction(transaction Transaction) error {
 	rb := &responseBuilder{
-		requestID:   requestID,
-		linkTracker: ra.GetProcess(p).(*peerLinkTracker),
+		requestID:   rs.requestID,
+		linkTracker: rs.linkTrackers.GetProcess(rs.p).(*peerLinkTracker),
 	}
 	err := transaction(rb)
-	ra.execute(p, rb.operations, rb.notifees)
+	rs.execute(rb.operations)
 	return err
 }
 
-func (ra *ResponseAssembler) execute(p peer.ID, operations []responseOperation, notifees []notifications.Notifee) {
+func (rs *responseStream) execute(operations []responseOperation) {
 	size := uint64(0)
 	for _, op := range operations {
 		size += op.size()
 	}
-	ra.peerHandler.AllocateAndBuildMessage(p, size, func(builder *gsmsg.Builder) {
+	if rs.isClosed() {
+		return
+	}
```

Comment:
So I think this (and the `isClosed()` 4 lines down) are the key to the "don't keep on queuing messages on an errored connection" part of the problem, is that correct? Basically making each new `Transaction()` on one of these errored connections a noop.

Reply:
Yep! That's the whole idea :)